de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/ResourceActivityQueuer.java

```diff
@@ -0,0 +1,85 @@
+package de.fu_berlin.inf.dpp.session.internal;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import de.fu_berlin.inf.dpp.activities.IActivity;
+import de.fu_berlin.inf.dpp.activities.IResourceActivity;
+import de.fu_berlin.inf.dpp.activities.SPath;
+import de.fu_berlin.inf.dpp.session.IActivityQueuer;
+
+/**
+ * This Resource based Queue gets a {@code Set} of {@link SPath}s, which should
+ * be queued and flushes them when the SPath entry was removed by
+ * {@link #disableQueuing(SPath)}.
+ * <p>
+ * To keep it simple, the execution order of queued activities stays unchanged.
+ * </p>
+ * TODO Discard obsolete Editor Events
+ */
+public class ResourceActivityQueuer implements IActivityQueuer {
+
+    private Set<SPath> queueRessources;
+    private List<IActivity> queueActivities = new LinkedList<IActivity>();
+
+    /**
+     * Takes a {@code Set} of {@link SPath}s for Queuing.
+     *
+     * @param queueRessources
+     */
+    public ResourceActivityQueuer(Set<SPath> queueRessources) {
+        ConcurrentHashMap<SPath, Boolean> map;
+        map = new ConcurrentHashMap<SPath, Boolean>(queueRessources.size() * 2);
+        this.queueRessources = Collections.newSetFromMap(map);
+        this.queueRessources.addAll(queueRessources);
+    }
+
+    /**
+     * Should be called after Resource is completely written to disk.
+     *
+     * @param resource
+     */
+    public void disableQueuing(SPath resource) {
+        queueRessources.remove(resource);
+    }
+
+    @Override
+    public List<IActivity> process(List<IActivity> activities) {
+        if (queueRessources.isEmpty())
+            return activities;
+
+        final List<IActivity> activitiesToExecute = new LinkedList<IActivity>();
+
+        /* check if queue activities are flushable */
+        for (IActivity activity : queueActivities) {
+            if (isAvailable(activity))
+                activitiesToExecute.add(activity);
+            else
+                break;
```

Comment:
Why do you use a `break` here? Isn't it possible that the resources for the first activities in the queue are not available, but the resources for activities later in the queue are?
You mention in the javadoc that activities are executed in the order they arrive in to keep it "simple". What purpose does it actually server to keep activities for files queued only because activities for another file could not be executed yet?

They way I understand it, activities for different files should not be able to influence each other, meaning it is only important that the execution order for each resource stays intact.
They could only influence whether the current state between the files is synchronized (when a method is moved from class A to class B while B is not loaded, the method is removed from A but not yet added to B), which should not matter since the "desynchronized" files are not present locally and will be synchronized right after they are loaded locally.

Reply:
Ok, I understand. Should be fine for now, but I would still prefer a better solution in the future. Could you add an issue for this?

Reply:
So basically it is not that instant. imagine a make a modification to a file that is written as last file on the receiving side.

as a side note:

file a, b, c

when i cycle through the editors in order c, a, b the result after the negotiation should be: i see file b open in the session overview, neither a or c. i guess that is not trivial if you want to release all activities at once for a resource that got available.

Reply:
Please think this in the context of instant session start in which this scenario is not going to happen / has no impact.
ISS doesn't use a fixed file order. Every opened and especially files opened during the negotiation are transmitted first. Even when you switch between opened files, the active file is transmitted next or probably is already transmitted as this is happening fast.

As a result (in normal / most cases) there is nearly no latency in applying of activities, because the transmission of even a big file like `SarosSession.java` (around 1100 LOC, compressed ~ 9 kB) is fast enough transmitted.

If you like to see a basic example of ISS in action [your'e welcome](https://stefaus.de/saros/instant.mkv)

---

My idea for the next iterations of this queuer is to queue things like open files, selected lines, ... different. Like if you already have a `open file activity` in the queue and get a new one discard the old (easier and probably sufficient, but possible side effects?! - there is some hack in the project based queuer...) or queue it for opening it in background.
But the impact of these cases is low, no data corruption and the next user input would already fix the wrong viewport...

So this queuer is only the first iteration and will get some improvements. 
After all ISS is for now an optional Transfer Type marked as experimental.

Reply:
This was a theoretical concern what could happen. If I would look at a more realistic scenario I would came to the conclusion that the major time of the negotiation is spent in calculating checksums and so in reality the difference between ISS and the old way is almost zero if you always reuse your projects.